### DISCUSSION TITLE
Budibase user data, app data security, release strategy for self host questions

### DISCUSSION THREAD

**Discussion by gmentzik** [Budibase/budibase#11118] [Help] [answered]
Hi devs,

I am trying to find some info regarding on how you secure that Budibase will not leak to internet sensitive user data, either this could be user mails or creds or data from application connected databases. I am asking because though I have managed to block most of communication from docker containers towards outside world I see that some things like getVersion are done from frontend. Also Budibase is allowed to downloads and  installs new docker image versions from admin backend probably through watchtower service which has direct access to docker daemon. Are there other case where Budibase is gaining direct access to system?
Could you help clarify and easy such security concerns?
Finally release strategy is a bit confusing. Currently last release note is on May but every week there is new release. Should someone update weekly? According to my experience frequent updates also bring bugs that are fixed at a next weekly release if reported.
Discussion is about self host version!

Thank you!

**Comment by tglynx** [comment]
Hi, 

i am not a budibase dev, however, as per my understanding if you do not expose any public screens a login is required to see your screens. Also, for using the api a token is required. I assume that all other requests to the bb application will get blocked. 

Maybe you could describe your concern or scenario in more detail? 

Regards,
Michael

**Reply by gmentzik** [reply]
Security concerns have to be addressed with a disclaimer that Budibase will not transmit any sensitive user or app data information towards internet either by sending data from FE or BE or exposing some API as you suggest. Is there such official documentation in any Budibase (Terms and Conditions doc or similar)? Any company security official would ask such info.

Secondly the release strategy is a bit confusing. While in github the last official version is of May 2023, new versions come up weekly. Self host version presents these weekly versions as the latest. Yet updating to them could cause issues with existing applications or backend functionality. So I would like to see somewhere a clear release strategy document that explains when an official release happens and when someone should update. Is there such a documentation?

**Reply by tglynx** [reply]
For the security part of your question: https://docs.budibase.com/docs/security

> The self-hosted version of budibase can be deployed entirely inside your own network, on your own servers. When running self-hosted, you have total control over how you want to secure your data, and it never has to leave your VPC.

If you only expose https to the bb frontend i would assume that the system can be considered safe.

For the release part: https://github.com/Budibase/budibase/releases
As far as i understand (and that may not be much) there are release notes for every release here.
Maybe i don't get what you mean by 'official release'...

What i am doing: 
I update my environment regularly. You can always go back to the previous version if you encounter any issues.
You could also setup a test environment to test newer releases with your apps first.

Thats just my thoughts...

**Reply by gmentzik** [reply]
Regarding security this does not cover us completely. What I understand from that paragraph is that Budibase does not take any responsibility for data leakage and you should keep it in an isolated VM if you use a self host. Example that I find weird is when you are in Budibase Admin portal, a getVersion request goes from FE to internet to get latest Budibase version. This cannot be stopped from backend security. I would like to know if other such things are happening either in BE or FE. 
Another example is that it sends statistics. I am not sure this is documented somewhere and how it can be stopped.
I am sure this is not the case and devs are taking all actions to guaranty that Budibase is secure.  So is such disclaimer available somewhere?

Regarding releases I know this page but I am trying to find somewhere which versions are considered safe and which under development. Self host fetches the latest in that page each time. Is this safe?

Finally you say you update your environment regularly and you can revert back to previous version if you encounter issues. Is there a restore for the whole Budibase or you just mean reverting the app. I am taking weekly backups of my system (images, volumes and with Budi cli script) and I can restore from that. Is there another way?

**Comment by shogunpurple** [comment, marked as answer]
@gmentzik 

First of all, it would be helpful to know how exactly you are using budibase and if it's in a security critical, company setting (internal company with private network) or if you are exposing your instance to the internet. We could guide you more on security when we know what you are planning to achieve with budibase.

> Regarding security this does not cover us completely. What I understand from that paragraph is that Budibase does not take any responsibility for data leakage and you should keep it in an isolated VM if you use a self host.

We build budibase with security at the forefront, and do our utmost to ensure that our code is secure, sensitive data is safe and vulnerabilities are addressed - including regular security audits from a 3rd party security company.

However, you have total control over your data in self host, and with that comes full responsibility. We cannot control how you use budibase when you self host and build apps, and if you happen to expose your data stored in your databases through your apps or misconfigure your security rules and are breached we cannot be responsible for "data leakage" that would occur in that scenario. If security is critical for a customer, or users are not comfortable with managing and provisioning their own security on self host - it's advisable to use budibase cloud or use a budibase plan that includes direct support. Happy to point you to the sales team if this is something that may be of interest.

You can read our privacy policy [here](https://budibase.com/privacy/) (mainly targeted at cloud and paid users)
Our [terms of service](https://budibase.com/terms-of-service/) may also contain relevant information

> Also Budibase is allowed to downloads and installs new docker image versions from admin backend probably through watchtower service which has direct access to docker daemon

Watchtower is optional. It's only a convenience for self hosters to be able to update through the UI. You can remove the watchtower container altogether and manage docker updates yourself if you wish.

> Example that I find weird is when you are in Budibase Admin portal, a getVersion request goes from FE to internet to get latest Budibase version. This cannot be stopped from backend security.

Yes, not every single call to the internet is going to be done from the backend in most apps if it's not necessary. If security is a high priority and you need to control/block **all** external network access for your users, it would be advisable to host your budibase instance internally and control access to your app via a private network, with network access control lists. The `getVersion` call is also optional - if you block that request budibase will still work - you will just have to keep an eye on the budibase github repo to know about the latest updates rather than in app. Some other calls done from the frontend include some analytics calls - which won't happen if you've opted out of analytics.


> Regarding releases I know this page but I am trying to find somewhere which versions are considered safe and which under development. Self host fetches the latest in that page each time. Is this safe?

Any versions pushed as github releases are "safe" and aren't under development. You will never pull an "in development" image from us by updating your budibase installation.

> Finally you say you update your environment regularly and you can revert back to previous version if you encounter issues. Is there a restore for the whole Budibase or you just mean reverting the app. I am taking weekly backups of my system (images, volumes and with Budi cli script) and I can restore from that. Is there another way?

A @tglynx mentions, if you are running budibase in production for mission critical apps, the best way to test new budibase versions before performing them on your production cluster is to use a test instance and verify. Another option of course is to backup using the CLI and roll back if something goes wrong as you are doing.
